optuna/study/_optimize.py

```diff
@@ -219,48 +220,35 @@ def _run_trial(
         state = TrialState.FAIL
         func_err = e
         func_err_fail_exc_info = sys.exc_info()
-    else:
-        # TODO(hvy): Avoid checking the values both here and inside `Study.tell`.
-        values, values_conversion_failure_message = _check_and_convert_to_values(
-            len(study.directions), value_or_values, trial.number
-        )
-        if values_conversion_failure_message is not None:
-            state = TrialState.FAIL
-        else:
-            state = TrialState.COMPLETE
 
     if study._storage.is_heartbeat_enabled():
         assert stop_event is not None
         assert thread is not None
         stop_event.set()
         thread.join()
 
-    # `Study.tell` may raise during trial post-processing.
-    try:
-        study.tell(trial, values=values, state=state)
-    except Exception:
-        raise
-    finally:
-        if state == TrialState.COMPLETE:
-            study._log_completed_trial(trial, cast(List[float], values))
-        elif state == TrialState.PRUNED:
-            _logger.info("Trial {} pruned. {}".format(trial.number, str(func_err)))
-        elif state == TrialState.FAIL:
-            if func_err is not None:
-                _logger.warning(
-                    "Trial {} failed because of the following error: {}".format(
-                        trial.number, repr(func_err)
-                    ),
-                    exc_info=func_err_fail_exc_info,
-                )
-            elif values_conversion_failure_message is not None:
-                _logger.warning(values_conversion_failure_message)
-            else:
-                assert False, "Should not reach."
-        else:
-            assert False, "Should not reach."
+    frozen_trial = study.tell(trial, values=value_or_values, state=state)
+
+    if frozen_trial.state == TrialState.COMPLETE:
+        study._log_completed_trial(trial, frozen_trial.values)
```

Comment:
How about updating `_log_completed_trial` to take the frozen trial, simplifying the internals?
```suggestion
        study._log_completed_trial(frozen_trial)
```

Reply:
Also, I think that the code would be cleaner if we avoid to use `trial` but instead `frozen_trial` from hereon.

Reply:
same as https://github.com/optuna/optuna/pull/3144#discussion_r780758665

Reply:
Looks great, thanks!